packages/block-editor/src/store/selectors.js

```diff
@@ -137,32 +140,76 @@ export function getBlockAttributes( state, clientId ) {
  * is not the block's registration settings, which must be retrieved from the
  * blocks module registration store.
  *
- * @param {Object} state    Editor state.
- * @param {string} clientId Block client ID.
+ * This selector handles two cases for different types of entities within the
+ * block editor:
+ * 1. A parent entity. For the parent entity, it wants to traverse through the
+ *    block tree, but stop when it reaches a different entity, like a template
+ *    part. In this scenario, withImmediateChildren is false deep into the block
+ *    tree. In other words, if an InnerBlock controller is reached, the selector
+ *    is only memoized based on the controller's attributes, NOT the block cache
+ *    state. The block cache state changes when the controlled inner blocks
+ *    change, which is not applicable to a parent entity, so we do not want to
+ *    trigger any updates.
+ * 2. A child controller. For the child controller, it needs to be aware of its
+ *    direct children. As a result, it is memoized on the block cache state,
+ *    since it does want to be aware of changes to its inner blocks. If it were
+ *    only memoized on attributes, it would not update as its inner blocks
+ *    change.
+ *
+ * In most cases, both of the above are true. For example, a template part acts
+ * as a "child" controller since it updates as its inner blocks change. But as
+ * it traverses into the block tree, it acts as a "parent", since
+ * withImmediateChildren will not be passed down as this selector recurses. So,
+ * as a controller, it can listen directly as its inner blocks change without
+ * seeing any changes inside of any controllers which may happen to exist in its
+ * inner blocks.
  *
+ * @param {Object}  state    Editor state.
+ * @param {string}  clientId Block client ID.
+ * @param {boolean} [withImmediateChildren] Use this parameter to ignore whether
+ *                                          inner blocks are controlled. Helpful
+ *                                          for the inner block component, which
+ *                                          needs to see its children, even though
+ *                                          they are controlled.
  * @return {Object} Parsed block object.
  */
 export const getBlock = createSelector(
-	( state, clientId ) => {
+	( state, clientId, withImmediateChildren = false ) => {
```

Comment:
I'm still interested in knowing in  which situations, we need this to  be true. I mean what if we don't have the flag, what would it break?

Reply:
`true` means "ignore inner blocks". We need it, so that parent entities don't become dirty from child entity edits. See the comment above and usage in the block editor provider and inner blocks.

Reply:
I mean where are the cases where we don't want to ignore nested entities?

Reply:
There are basically two separate situations we are thinking about here. 

When you call `getBlock` from the inner block component, the component is both an inner block controller, and it also needs to be aware of changes happening inside of itself. This effectively lets us get around the check for "controlled inner blocks" a few lines below this one. 

If this value is true, then we do return `innerBlocks` despite the fact that the current block is an inner block controller. This is _only_ true when calling `getBlock` from the context of the `InnerBlocks` component. Since it is true, `getBlock` returns the immediate inner blocks of the given block, _even though_ the current block is an inner block controller.

As we recurse down through the tree, though, this value is false. As a result, any other inner blocks inside of an inner block controller are _not_ returned. This means that we can support getting inner blocks at the root level of an inner block controller, and also support not getting the inner blocks of any children inner block controllers.

If we kept this value always true, then inner blocks would always be returned despite the inner block controllers. If we keep the value always false, then inner blocks will never be returned at the top level of the inner block controller. So we do both: it is true at the top level of the inner block controller (and so the template part top-level blocks show up in the editor), but false as we recurse further into the block tree. (so that any nested template parts are not considered to be part of their parent)

Does that make sense?

Reply:
`getBlock( rootClientId )` to get `.innerBlocks` to render in the inner blocks component.

Reply:
I changed function such that it never returns inner blocks if they are controlled